state_test.go

```diff
@@ -24,68 +71,307 @@ func TestState_FaultyNodes(t *testing.T) {
+	maxRound, found = s.maxRound()
+	assert.Equal(t, uint64(0), maxRound)
+	assert.Equal(t, false, found)
 }
 
+func TestState_AddRoundMessage(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"validator_1", "validator_2"}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	roundMessageSize := s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_Commit, 0))
+	assert.Equal(t, 0, roundMessageSize)
+	assert.Equal(t, 0, len(s.roundMessages))
+
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 0))
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 1))
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 2))
+
+	roundMessageSize = s.AddRoundMessage(createMessage(validatorIds[1], MessageReq_RoundChange, 2))
+	assert.Equal(t, 2, roundMessageSize)
+
+	s.AddRoundMessage(createMessage(validatorIds[1], MessageReq_RoundChange, 3))
+	assert.Equal(t, 4, len(s.roundMessages))
+
+	assert.Empty(t, s.prepared)
+	assert.Empty(t, s.committed)
+}
+
+func TestState_addPrepared(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"validator_1", "validator_2"}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	s.addPrepared(createMessage(validatorIds[0], MessageReq_Commit))
+	assert.Equal(t, 0, len(s.prepared))
+
+	s.addPrepared(createMessage(validatorIds[0], MessageReq_Prepare))
+	s.addPrepared(createMessage(validatorIds[1], MessageReq_Prepare))
+
+	assert.Equal(t, len(validatorIds), len(s.prepared))
+	assert.Empty(t, s.committed)
+	assert.Empty(t, s.roundMessages)
+}
+
+func TestState_addCommitted(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"validator_1", "validator_2"}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	s.addCommitted(createMessage(validatorIds[0], MessageReq_Prepare))
+	assert.Empty(t, 0, s.committed)
+
+	s.addCommitted(createMessage(validatorIds[0], MessageReq_Commit))
+	s.addCommitted(createMessage(validatorIds[1], MessageReq_Commit))
+
+	assert.Equal(t, len(validatorIds), len(s.committed))
+	assert.Empty(t, s.prepared)
+	assert.Empty(t, s.roundMessages)
+}
+
+func TestState_Copy(t *testing.T) {
+	originalMsg := createMessage("validator_1", MessageReq_Preprepare, 0)
+	copyMsg := originalMsg.Copy()
+	assert.NotSame(t, originalMsg, copyMsg)
+	assert.Equal(t, originalMsg, copyMsg)
+}
+
+func TestState_Lock_Unlock(t *testing.T) {
+	s := newState()
+	proposalData := make([]byte, 2)
+	mrand.Read(proposalData)
+	s.proposal = &Proposal{
+		Data: proposalData,
+		Time: time.Now(),
+	}
+	s.lock()
+	assert.True(t, s.locked)
+	assert.NotNil(t, s.proposal)
+
+	s.unlock()
+	assert.False(t, s.locked)
+	assert.Nil(t, s.proposal)
+}
+
+func TestState_GetSequence(t *testing.T) {
+	s := newState()
+	s.view = &View{Sequence: 3, Round: 0}
+	assert.True(t, s.GetSequence() == 3)
+}
+
+func TestState_getCommittedSeals(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"A", "B"}
+	s.validators = newMockValidatorSet(validatorIds)
```

Comment:
Do not be afraid to create a pool object here only to get the validator set. It is convenient.

Reply:
Have changed it, although it seems like it is slower than creating mock validator set, as I have pointed out in the https://github.com/0xPolygon/pbft-consensus/pull/12#discussion_r786076784.